# Step 18. Federation, one module at a time

Step 07 proved that pooling `N, S, Q, G` reproduces the pooled correlation matrix exactly, and named
the blocker: `rank(G) = min(n, p)`, so a Gram matrix from a panel wider than the cohort can be
inverted back toward individual rows. The release rule is `p < n`.

An earlier version of this analysis concluded federation was impossible here. That was wrong. It
tested each cohort's whole panel as a single Gram, 430 to 771 proteins against n = 86, which
nothing in the method requires. Modules are small. Released one at a time, and counted in
proteins rather than probes, they clear the rule with room to spare:

| cohort | trait-associated modules | surviving step 12's 1% size rule | individually releasable, in proteins |
|---|---|---|---|
| A | 9 | 6 | all: 23, 28, 32, 32, 39 |
| B | 2 | 1 | 22 |
| C | 1 | 0 | none survive the size rule; the one module federated from C here is its renal module, 38 |

Every module released below clears `p < n = 86` with room to spare, the largest at 39 proteins.

Counting proteins rather than probes matters exactly here. The interferon module is 27 probes but
**23 proteins**, because ISG15 and STAT1 each carry two SOMAmers; cohort A's `royalblue` is 34 probes
but **28 proteins**. Counting probes inflates `p` precisely where modules are smallest and the rule is
easiest to satisfy.

## One module per run

This notebook federates a single named module and releases nothing else. Nine separate Grams
totalling 222 proteins is not obviously the same disclosure as one 222-protein Gram, each block
fixes its own proteins only up to an unknown rotation, and the blocks cannot be aligned to each
other, but that is an argument, not a citation. It is an open privacy question, and this
notebook is built so it never has to be answered: one module, one release, one decision.

## Every cohort now contributes, and B originates one

Under the published parameters B has two trait-associated modules, one of which, `cyan`, survives the
size rule at 22 proteins, so B defines a module as well as shipping sufficient statistics for modules
defined elsewhere. Cohort C is the asymmetric one now: its single trait-associated module covers 5.9%
of the panel and is excluded, so C originates nothing through that route. What each cohort can
contribute is an outcome of what it contains, and it is stated rather than engineered away.

In [1]:
source("../src/paths.R")
suppressMessages(library(WGCNA)); options(stringsAsFactors=FALSE); set.seed(SEED)
SITES <- c("A","B","C")
W  <- lapply(SITES, function(s) readRDS(art("wgcna_%s.rds", s))); names(W) <- SITES
N   <- sapply(W, function(w) nrow(w$X)); N_MIN <- min(N)
spec <- read_traits()
build_traits <- function(m, spec){ out<-data.frame(row.names=rownames(m))
  for (i in seq_len(nrow(spec))){ v<-m[[spec$source_column[i]]]
    out[[spec$name[i]]] <- switch(spec$type[i], numeric=as.numeric(as.character(v)),
      binary=as.numeric(v=="Positive"),
      ordinal={lvl<-unique(v[!is.na(v)&v!=""]); lvl<-lvl[order(as.numeric(sub("-.*","",lvl)))]
               as.integer(factor(v,levels=lvl,ordered=TRUE))}) }; out }
TR <- lapply(SITES, function(s) build_traits(W[[s]]$meta, spec)); names(TR)<-SITES

### The algebra, unchanged from step 07

`suff()` and `pooled_cor()` are step 07's, including the clamp added in step 10: rebuilding `r` from
sufficient statistics overshoots ±1 by ~1e-16 on near-perfect pairs, and weighted gene co-expression network analysis (WGCNA)'s `checkAdjMat`
rejects that.

`federate_one()` refuses any module at or above `n`. It does not trim a module to make it fit, trimming would be choosing which proteins to drop in order to pass a privacy rule, which is the
wrong thing to optimise.

In [2]:
suff <- function(M){ M<-as.matrix(M)
  list(N=crossprod(!is.na(M)), S=crossprod(!is.na(M), replace(M,is.na(M),0)),
       Q=crossprod(!is.na(M), replace(M,is.na(M),0)^2), G=crossprod(replace(M,is.na(M),0))) }
pooled_cor <- function(st){ n<-st$N; s<-st$S; g<-st$G; q<-st$Q
  cov <- g/n - (s*t(s))/(n*n); sv <- sqrt(q/n - (s/n)^2)
  pmin(pmax(cov/(sv*t(sv)), -1), 1) }

federate_one <- function(ORIGIN, MODULE){
  g <- colnames(W[[ORIGIN]]$X)[W[[ORIGIN]]$mods == MODULE]
  if (!length(g)) {          # the colour does not exist in this cohort's fit
    cat(sprintf("\n=== module '%s' at cohort %s : NOT PRESENT in this fit, skipped ===\n",
                MODULE, ORIGIN)); return(invisible(NULL)) }
  np <- n_proteins(g)
  cat(sprintf("\n=== module '%s', defined at cohort %s : %s ===\n", MODULE, ORIGIN, size_str(g)))
  if (np >= N_MIN){
    cat(sprintf("  REFUSED: %d proteins is not < n = %d. Not trimming a module to make it fit.\n",
                np, N_MIN)); return(invisible(NULL)) }
  cat(sprintf("  eligible: %d proteins < n = %d\n", np, N_MIN))
  st <- Reduce(function(a,b) Map(`+`,a,b), lapply(SITES, function(s) suff(W[[s]]$X[,g,drop=FALSE])))
  R  <- pooled_cor(st)
  Rd <- cor(do.call(rbind, lapply(SITES, function(s) W[[s]]$X[,g,drop=FALSE])))
  ex <- max(abs(R-Rd)); cat(sprintf("  exactness gate: max|federated - pooled raw| = %.1e\n", ex))
  stopifnot(ex < 1e-9)
  n_tot <- st$N[1,1]; mu <- st$S[1,]/n_tot; sdv <- sqrt(st$Q[1,]/n_tot - mu^2)
  v <- eigen(R, symmetric=TRUE)$vectors[,1]; if (sum(v) < 0) v <- -v; names(v) <- g
  score <- lapply(SITES, function(s){
    Z <- scale(W[[s]]$X[,g,drop=FALSE], center=mu, scale=sdv); as.vector(Z %*% v) })
  names(score) <- SITES
  me <- moduleEigengenes(W[[ORIGIN]]$X[,g,drop=FALSE], rep("m", length(g)))$eigengenes[[1]]
  rt <- abs(cor(score[[ORIGIN]], me))
  cat(sprintf("  round-trip vs local moduleEigengenes at %s: r = %.4f\n", ORIGIN, rt))
  coh <- sapply(SITES, function(s){ C<-cor(W[[s]]$X[,g,drop=FALSE]); mean(C[upper.tri(C)]) })
  cat("  mean within-module correlation per site: ",
      paste(sprintf("%s %.2f", SITES, coh), collapse="  "), "\n")
  hits <- do.call(rbind, lapply(SITES, function(s){
    tr <- TR[[s]][rownames(W[[s]]$X),,drop=FALSE]
    # Spearman and the per-pair n, as in step 04, which explains both.
    r <- cor(score[[s]], tr, method="spearman", use="pairwise.complete.obs")[1,]
    nn <- colSums(!is.na(as.matrix(tr)) & !is.na(c(score[[s]])))
    p <- 2*pt(-abs(r*sqrt((nn-2)/(1-r^2))), nn-2)
    q <- p.adjust(p,"BH")
    k <- which(q<0.05)
    if (!length(k)) return(NULL)
    data.frame(site=s, trait=names(r)[k], r=round(r[k],2), q=signif(q[k],2)) }))
  cat("  trait associations at FDR 5%:\n")
  if (is.null(hits)) cat("    none\n") else print(hits[order(hits$site,-abs(hits$r)),], row.names=FALSE)
  invisible(list(module=MODULE, origin=ORIGIN, proteins=g, R=R, v=v, score=score,
                 exact=ex, roundtrip=rt, coherence=coh, hits=hits))
}

### The demonstration

The modules are **derived, not named**: the interferon module from step 02b, plus every
trait-associated module that survived step 12's size exclusion, in each cohort. That list used to be
five colours written out by hand, and when the network type changed from signed to unsigned every one
of them ceased to exist and this notebook failed. A WGCNA colour is meaningful only inside one fit, so
it can never be a literal here.

Seven modules result: five defined at A, one at B and one at C. The refusal path is still reachable —
`federate_one()` prints `REFUSED` for any module of `n` proteins or more, and skips a colour absent
from a cohort's fit — but on this fit nothing triggers it, because every surviving module is small.

In [3]:
# The modules to federate are DERIVED, never named by colour. The list here used to
# be five literal colours from one particular fit; when the network became unsigned
# they all ceased to exist and this notebook failed. What it federates now comes from
# the two decisions already recorded upstream: the interferon module named in step
# 02b, and every trait-associated module that survived step 12's size exclusion.
IFN_MOD <- readRDS(art("ifn_A.rds"))$module
P12  <- readRDS(art("step12_panels.rds"))$D
DEMO <- c(list(c("A", IFN_MOD)), unlist(lapply(SITES, function(s) {
  k <- P12[[paste(s, "all15")]]$keep
  lapply(setdiff(k, if (s == "A") IFN_MOD else character(0)), function(m) c(s, m))
}), recursive = FALSE))
cat("modules to federate, from step 02b and step 12:\n")
for (d in DEMO) cat(sprintf("   cohort %s / %s\n", d[1], d[2]))
res <- lapply(DEMO, function(x) federate_one(x[1], x[2]))
names(res) <- sapply(DEMO, function(x) paste(x, collapse="/"))
saveRDS(res[!sapply(res, is.null)], art("step18_federated_modules.rds"))

modules to federate, from step 02b and step 12:


   cohort A / skyblue
   cohort A / darkgreen
   cohort A / darkred
   cohort A / greenyellow
   cohort A / royalblue
   cohort B / cyan
   cohort C / royalblue



=== module 'skyblue', defined at cohort A : 27 probes (23 proteins) ===
  eligible: 23 proteins < n = 86
  exactness gate: max|federated - pooled raw| = 9.1e-13
  round-trip vs local moduleEigengenes at A: r = 0.9993
  mean within-module correlation per site:  A 0.22  B 0.18  C 0.17 
  trait associations at FDR 5%:
 site            trait     r       q
    A        Sm_status  0.53 2.0e-06
    A     Ro_60_status  0.46 6.4e-05
    A     dsDNA_status  0.42 4.1e-04
    A   Duration_years -0.41 4.1e-04
    A         Age_band -0.40 4.1e-04
    A         C3_level -0.39 4.1e-04
    A     RNP_A_status  0.39 4.1e-04
    A        SLEDAI_2K  0.34 2.2e-03
    A     Ro_52_status  0.33 3.3e-03
    A Lymphocyte_count -0.32 3.7e-03
    A    RNP_68_status  0.32 3.3e-03
    A         C4_level -0.29 8.0e-03
    A       Creatinine -0.22 4.2e-02
    B     Ro_60_status  0.34 1.2e-02
    B         C4_level -0.32 1.2e-02
    B Lymphocyte_count -0.32 1.2e-02
    B    RNP_68_status  0.31 1.3e-02
    B         C3

## What federation buys

**The arithmetic is exact.** Across all seven modules the gate `max|federated - pooled raw|` runs from
**7.0e-13 to 1.5e-12**, which is floating-point noise. Pooling `N, S, Q, G` reproduces the pooled
correlation matrix, as step 07 proved it must.

**Two modules carry a signal in all three cohorts, and they are the paper's two.**

*Interferon, `skyblue`, defined at A, 27 probes (23 proteins).* Associations at FDR 5% in **A (13
traits), B (10) and C (4)**, and the antibodies are the same ones in each:

| antibody | A | B | C |
|---|---|---|---|
| anti-Sm | +0.53, q 2.0e-06 | +0.27, q 0.018 | +0.43, q 3.8e-04 |
| anti-Ro-60 | +0.46, q 6.4e-05 | +0.34, q 0.012 | — |
| anti-RNP-A | +0.39, q 4.1e-04 | +0.24, q 0.042 | +0.44, q 3.8e-04 |
| anti-RNP-68 | +0.32, q 3.3e-03 | +0.31, q 0.013 | +0.34, q 9.2e-03 |

This is the source paper's own finding, reached here independently on three cohorts of ~87: Leung et
al. report that "positivity for antibodies targeting RNA-binding proteins (anti-Sm, anti-Ro-60,
anti-RNP68, anti-RNP-A) was associated with increased IFN-stimulation". All four, all positive.

*Renal, `royalblue`, defined at A, 34 probes (28 proteins).* Associates with **uPCR in all three**:
A +0.41 (q 0.0011), B +0.37 (q 0.0066), C +0.43 (q 0.00054). Its within-module coherence is close
across cohorts too, 0.38 / 0.26 / 0.27.

**Cohort B does originate a module now.** `cyan`, 26 probes (22 proteins), is B's one module that
survives the size rule. It is the most internally coherent module in the set — 0.49 / 0.58 / 0.52
across the three cohorts, higher at A and C than several modules defined *at* A — yet its only
association is `Age_band` at B. Structure without phenotype, and worth saying rather than dropping.

**Coherence is where non-replication shows first.** `darkgreen` (A, 32 proteins) has within-module
correlation **0.53 at A but 0.04 at B and 0.04 at C**: the proteins move together at A and nowhere
else. `darkred` is 0.01 / 0.04 / 0.05 — barely coherent anywhere, including where it was defined.
`royalblue` defined at **C** is 0.50 at C and 0.09 at both others. Step 20 tests this properly with
permutations; the coherence column is the cheap early warning.

**Round-trip is 0.9913 to 0.9993, not 1.0000, and that is correct.** The federated loadings centre and
scale with the pooled per-protein mean and sd, while a local `moduleEigengenes` fit uses that cohort's
own. The gap is the difference between the two scalings, not a refit. Step 08's projection gate
returns exactly 1.0000 because it uses one cohort's scaling throughout.

**What this does not do.** Discovery still needs the full panel and therefore one site: the module has
to exist before it can be released. This is federated confirmation, step 07's second option, with the
arithmetic now performed rather than described. Step 19 measures whether the federated loadings add
anything over simply sharing the protein list, and across these seven modules the answer is mostly no.

## Where this was run

Rendered notebooks are committed, so each records the machine, the R and the
package versions that produced its output.


In [4]:
# Provenance. The rendered notebook is committed, so it records where it ran.
# The cell is identical in every notebook and is explained in step 00.
si <- Sys.info()
cat("run on      :", si[["nodename"]], "(", si[["sysname"]], si[["release"]], ")\n")
cat("date        :", format(Sys.time(), "%Y-%m-%d %H:%M %Z"), "\n")
cat("R           :", R.version.string, "|", R.version$platform, "\n")
cat("R comes from:", R.home(), "\n")
cat("packages    :\n")
for (p in c("WGCNA", "ComplexHeatmap", "circlize", "ReactomePA", "reactome.db",
            "org.Hs.eg.db", "sva", "VarSelLCM"))
  cat(sprintf("   %-16s %s\n", p,
              tryCatch(as.character(packageVersion(p)), error = function(e) "not installed")))

run on      : Annes-MacBook-Pro-193.local ( Darwin 27.0.0 )


date        : 2026-09-28 01:04 EDT 


R           : R version 4.4.3 (2025-02-28) | x86_64-apple-darwin13.4.0 


R comes from: /Users/adeslatt/miniforge3/envs/endotypes-proteomics/lib/R 


packages    :


   WGCNA            1.74
   ComplexHeatmap   2.22.0
   circlize         0.4.18
   ReactomePA       1.50.0
   reactome.db      1.89.0
   org.Hs.eg.db     3.20.0
   sva              3.54.0
   VarSelLCM        2.1.3.2
